gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java

```diff
@@ -363,11 +357,10 @@ private static void put(ByteString source, int offset, int size, ByteBuffer dest
 
   private int readBufferedContentInto(ByteBuffer byteBuffer) {
     // Handle skipping forward through the buffer for a seek.
-    long bufferSkip =
-        min(bufferedContent.size() - bufferedContentReadOffset, bytesToSkipBeforeReading);
+    long bytesToSkip = positionForNextRead - positionInGrpcStream;
+    long bufferSkip = min(bufferedContent.size() - bufferedContentReadOffset, bytesToSkip);
```

Comment:
while we are at it (simplifying the logic) we can look at getting rid of `bufferedContentReadOffset` variable as well. We can ensure that `bufferedContent` always has relevant data from offset 0. Else we can  do 
```
   bufferedContent = bufferedContent.substring(skipBytes)
```
This reduces the complexity of maintaining both `bufferedContent` and `bytesToSkipBeforeReading`